### ISSUE TITLE
URLCanonicalizer used to transform outgoing URLs

### ISSUE THREAD

**Issue by Phil4** [yasserg/crawler4j#123]
When using Crawl4j, I often visit websites that do not accept their URLs are encoded according to RFC 3986. Especially when it comes to replace a '+' with '%2B' on query params. I end up crawling an error page instead of the actual page.

Now, those websites have certainly a poor URL management but I don't have control over them... So I think it would be great to handle those cases.

Correct me if I'm wrong but I think the main use of canonicalizing URLs is to avoid to crawl again the same URL twice. This could maybe be used only internally (to check if an URL has already been visited) without modifying the actual crawled URL (which has no reason to differ) ?

**Comment by Anebrithien** [comment]
In my case, the seed page said 

```html
  <META http-equiv=Content-Type content="text/html; charset=gb2312">
```
and the href displayed as 

```html
  <a href="3.asp?DengJh=洪1700718" target="_blank">
```

Then `URLCanonicalizer#percentEncodeRfc3986(String)` transform query params to `DengJh=%E6%B4%AA1700718`,  caught HTTP 404 error as last.

The real link in this website should be `DengJh=%BA%E91700718`, we should encode `洪1700718` in `gb2312` instead of `UTF-8`.

Here are some test code:

```java
  System.out.println(new String(new byte[]{(byte) 0xBA, (byte) 0xE9}, Charset.forName("GBK"))); 
  System.out.println(new String(new byte[]{(byte) 0xE6, (byte) 0xB4, (byte) 0xAA}, Charset.forName("UTF-8")));
  System.out.println("target= " + URLEncoder.encode("洪1700718", "GBK"));
```

**Comment by s17t** [comment]
Fixed in #272
